core/tooltip.js

```diff
@@ -234,235 +32,421 @@ const getTargetObject = function(obj) {
+    this.element = null;
+    this.poisonedElement = null;
+    this.hide();
+  }
 
-/**
- * Create the tooltip and show it.
- */
-const show = function() {
-  if (blocked) {
-    // Someone doesn't want us to show tooltips.
-    return;
+  /**
+   * Hide the tooltip.
+   * @alias Blockly.Tooltip.hide
+   */
+  hide() {
+    if (this.visible) {
+      this.visible = false;
+      if (this.DIV) {
+        this.DIV.style.display = 'none';
+      }
+    }
+    if (this.showPid) {
+      clearTimeout(this.showPid);
+    }
   }
-  poisonedElement = element;
-  if (!DIV) {
-    return;
+
+  /**
+   * Hide any in-progress tooltips and block showing new tooltips until the next
+   * call to unblock().
+   * @alias Blockly.Tooltip.block
+   * @package
+   */
+  block() {
+    this.hide();
+    this.blocked = true;
   }
-  // Erase all existing text.
-  DIV.textContent = '';
-  let tip = getTooltipOfObject(element);
-  tip = blocklyString.wrap(tip, LIMIT);
-  // Create new text, line by line.
-  const lines = tip.split('\n');
-  for (let i = 0; i < lines.length; i++) {
-    const div = document.createElement('div');
-    div.appendChild(document.createTextNode(lines[i]));
-    DIV.appendChild(div);
+
+  /**
+   * Unblock tooltips: allow them to be scheduled and shown according to their
+   * own logic.
+   * @alias Blockly.Tooltip.unblock
+   * @package
+   */
+  unblock() {
+    this.blocked = false;
   }
-  const rtl = /** @type {{RTL: boolean}} */ (element).RTL;
-  const windowWidth = document.documentElement.clientWidth;
-  const windowHeight = document.documentElement.clientHeight;
-  // Display the tooltip.
-  DIV.style.direction = rtl ? 'rtl' : 'ltr';
-  DIV.style.display = 'block';
-  visible = true;
-  // Move the tooltip to just below the cursor.
-  let anchorX = lastX;
-  if (rtl) {
-    anchorX -= OFFSET_X + DIV.offsetWidth;
-  } else {
-    anchorX += OFFSET_X;
+
+  /**
+   * Renders the tooltip content into the tooltip div.
+   * @protected
+   */
+  renderContent() {
+    let tip = this.getTooltipOfObject(this.element);
+    tip = blocklyString.wrap(tip, this.LIMIT);
+    // Create new text, line by line.
+    const lines = tip.split('\n');
+    for (let i = 0; i < lines.length; i++) {
+      const div = document.createElement('div');
+      div.appendChild(document.createTextNode(lines[i]));
+      this.DIV.appendChild(div);
+    }
   }
-  let anchorY = lastY + OFFSET_Y;
 
-  if (anchorY + DIV.offsetHeight > windowHeight + window.scrollY) {
-    // Falling off the bottom of the screen; shift the tooltip up.
-    anchorY -= DIV.offsetHeight + 2 * OFFSET_Y;
+  /**
+   * Gets the coordinates for the tooltip div, taking into account the edges of
+   * the screen to prevent showing the tooltip offscreen.
+   * @param {boolean} rtl True if the tooltip should be in right-to-left layout.
+   * @returns {{x: number, y: number}} Coordinates at which the tooltip div
+   *     should be placed.
+   * @protected
+   */
+  getPosition(rtl) {
```

Comment:
Opted not to do this since it's not overrideable and not even exported, so now the only reason it's broken apart is for readability and I don't think breaking it further enhances readability but lmk if you disagree!